cardano-api-test/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,263 @@
+
+
+-- * Ugrading scripts
+--
+-- The instruction set from V1 may be used as V2,
+-- and we can't determine the language version from the transaction.
+-- When the user uses V1 instructions,
+-- we don't know if they used the V1 tag or the V2 tag.
+-- So it's safe to drop this information.
+
+upgradeScriptInEra :: ScriptInEra era -> ScriptInEra era
+upgradeScriptInEra = \case
+  ScriptInEra SimpleScriptV1InAllegra script ->
+    ScriptInEra SimpleScriptV2InAllegra $ upgradeScript script
+  ScriptInEra SimpleScriptV1InMary script ->
+    ScriptInEra SimpleScriptV2InMary $ upgradeScript script
+  ScriptInEra SimpleScriptV1InAlonzo script ->
+    ScriptInEra SimpleScriptV2InAlonzo $ upgradeScript script
+  other -> other
+
+upgradeScript :: Script SimpleScriptV1 -> Script SimpleScriptV2
+upgradeScript (SimpleScript SimpleScriptV1 script) =
+  SimpleScript SimpleScriptV2 $ upgradeSimpleScript script
+
+upgradeSimpleScript ::
+  SimpleScript SimpleScriptV1 -> SimpleScript SimpleScriptV2
+upgradeSimpleScript = \case
+  RequireSignature hash -> RequireSignature hash
+  RequireAllOf scripts  -> RequireAllOf $ map upgradeSimpleScript scripts
+  RequireAnyOf scripts  -> RequireAnyOf $ map upgradeSimpleScript scripts
+  RequireMOf n scripts  -> RequireMOf n $ map upgradeSimpleScript scripts
+
+
+review :: TxBodyContent BuildTx era -> TxBodyContent ViewTx era
+review TxBodyContent{..} =
+  TxBodyContent
+    { txCertificates    = reviewCertificates txCertificates
+    , txExtraScriptData = ViewTx
+    , txIns             = map reviewTxIn     txIns
+    , txMintValue       = reviewMintValue    txMintValue
+    , txProtocolParams  = ViewTx
+    , txWithdrawals     = reviewWithdrawals  txWithdrawals
+    , ..
+    }
+
+reviewTxIn ::
+  (TxIn, BuildTxWith BuildTx (Witness WitCtxTxIn era)) ->
+  (TxIn, BuildTxWith ViewTx  (Witness WitCtxTxIn era))
+reviewTxIn = second $ const ViewTx
+
+reviewWithdrawals :: TxWithdrawals BuildTx era -> TxWithdrawals ViewTx era
+reviewWithdrawals = \case
+  TxWithdrawalsNone                 -> TxWithdrawalsNone
+  TxWithdrawals support withdrawals ->
+    TxWithdrawals
+      support
+      [(address, amount, ViewTx) | (address, amount, _) <- withdrawals]
+
+reviewCertificates :: TxCertificates BuildTx era -> TxCertificates ViewTx era
+reviewCertificates = \case
+  TxCertificatesNone                    -> TxCertificatesNone
+  TxCertificates support certificates _ ->
+    TxCertificates support certificates ViewTx
+
+reviewMintValue :: TxMintValue BuildTx era -> TxMintValue ViewTx era
+reviewMintValue = \case
+  TxMintNone                  -> TxMintNone
+  TxMintValue support value _ -> TxMintValue support value ViewTx
+
+
+test_roundtrip_TxBody_get_make :: [TestTree]
+test_roundtrip_TxBody_get_make =
+  [ testProperty (show era) $
+    property $ do
+      txbody <- forAll $ genTxBody era
+      tripping
+        txbody
+        (\(TxBody content) -> content)
+        (makeTransactionBody . rebuildBodyContent)
```

Comment:
When you document this property, make it clear why we need `genTxBody` and `rebuildBodyContent` rather than `genTxBody`. It's not clear on first reading, and `rebuildBodyContent` of course adds quite a bit of code, so there must be a good reason!

Reply:
This is a roundtrip path. On the way back, we start with `TxBodyContent ViewTx`, and to get back to `TxBody` we have to build it again.

Should I put this in the code?